### PR TITLE
Minimal ILGenerator implementation for new AssemblyBuilder.Save

### PR DESCRIPTION
**Pull request by buyaa-n**
Basically, replacing the ILGenerator.Emit overloads implementaiton with `InstructionEncoder` which will be written into the file when saving the methods in a type using [MethodBodyStreamEncoder](https://learn.microsoft.com/dotnet/api/system.reflection.metadata.ecma335.methodbodystreamencoder)

Contributes to https://github.com/dotnet/runtime/issues/83988

### CODE REVIEW

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        private const int DefaultSize = 16;
+        private readonly MethodBuilder _methodBuilder;
+        private readonly BlobBuilder _builder;
+        private readonly InstructionEncoder _il;
+        private bool _hasDynamicStackAllocation;
+
+        internal ILGeneratorImpl(MethodBuilder methodBuilder, int size)
+        {
+            _methodBuilder = methodBuilder;
+            _builder = new BlobBuilder(Math.Max(size, DefaultSize));
```
Comment (danmoseley) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:23]: Should it throw on size < 1?

Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:23]: For compat with the existing runtime ILGenerator:
https://github.com/dotnet/runtime/blob/a590cb4cfb9f1a66c043476695fd0e79835842eb/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/RuntimeILGenerator.cs#L90

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs [hunk 2]
```diff
@@ -180,11 +180,18 @@ internal void AppendMetadata()
...
         }
 
-        private void WriteMethods(TypeBuilderImpl typeBuilder, List<GenericTypeParameterBuilderImpl> genericParams)
+        private void WriteMethods(TypeBuilderImpl typeBuilder, List<GenericTypeParameterBuilderImpl> genericParams, MethodBodyStreamEncoder methodBodyEncoder)
         {
             foreach (MethodBuilderImpl method in typeBuilder._methodDefinitions)
             {
-                MethodDefinitionHandle methodHandle = AddMethodDefinition(method, method.GetMethodSignatureBlob(), _nextParameterRowId);
+                int offset = -1;
+                ILGeneratorImpl? il = method.ILGeneratorImpl;
```
Comment (steveharter) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:188]: Check for `!method.IsAbstract` before setting variable for `method.ILGeneratorImpl;`?

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs (hunk 1 above)
Comment (jeffhandley) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:23]: That might be worth a comment. It looks like the scenario is covered in a unit test, but the test could use a comment too.

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs [hunk 3]
```diff
@@ -230,6 +237,14 @@ private void WriteMethods(TypeBuilderImpl typeBuilder, List<GenericTypeParameter
             }
         }
 
+        private static int AddMethodBody(MethodBuilderImpl method, ILGeneratorImpl il, MethodBodyStreamEncoder methodBodyEncoder) =>
+            methodBodyEncoder.AddMethodBody(
+                instructionEncoder: il.Instructions,
+                maxStack: 8, // TODO
```
Comment (steveharter) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:243]: Per offline discussion, we should add an implementation to calculate this with this PR.

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs [hunk 4]
```diff
@@ -98,7 +111,23 @@ protected override ParameterBuilder DefineParameterCore(int position, ParameterA
...
+        protected override ILGenerator GetILGeneratorCore(int size)
+        {
+            if (IsGenericMethod && !IsGenericMethodDefinition)
+            {
+                throw new InvalidOperationException();
+            }
+
+            if ((_methodImplFlags & MethodImplAttributes.CodeTypeMask) != MethodImplAttributes.IL ||
+                (_methodImplFlags & MethodImplAttributes.Unmanaged) != 0 ||
+                (_attributes & MethodAttributes.PinvokeImpl) != 0)
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs:123]: The CoreLib implementation also has a check for DllImport here. Is it intentionally omitted?

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs (hunk 2 above)
Comment (jkotas) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:188]: Do we even need `method.IsAbstract` check here? It does not look like right to throw away the IL body.

Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:188]: Good point, the runtime implementation throws while populating the method bodies on `TypeBuilder.CreateTypeInfo()`:
https://github.com/dotnet/runtime/blob/087ca4662309f3db40b3096cbf89853d9c292e16/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/RuntimeTypeBuilder.cs#L1696
As we are not populating the `TypeInfo` similarly, probably we should check it and throw in `MethodBuilder.GetILGenerator()`?

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs (hunk 4 above)
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs:123]: We are setting the attribute when `DllImportAttribute` is set, no need to keep another boolean flag for this https://github.com/dotnet/runtime/blob/42a6776e6cb1569fe277181f424ccbbe243cf386/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs#L111-L114

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs (hunk 2 above)
Comment (jkotas) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:188]: I agree.

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:243]: If we can set `maxstack == the number of opcodes` for now that should protect us against any `InvalidProgramException` thrown by the JIT etc.

Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:243]: I can do that until I have got a better handling

**Review by steveharter — APPROVED** [review]
LGTM pending the discussion on a value for `maxstack` that shouldn't result in an exception if exceeded.

**Comment by buyaa-n** [comment]
The failures not related and known

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs [hunk 5]
```diff
@@ -46,6 +49,10 @@ public override void Emit(OpCode opcode)
                 _hasDynamicStackAllocation = true;
             }
             _il.OpCode((ILOpCode)opcode.Value);
+
+            // TODO for now only count the Opcodes emitted, in order to correctly we might need to make below API public
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:53]:
```suggestion
            // TODO: for now only count the Opcodes emitted, in order to calculate it correctly we might need to make internal Opcode APIs public
```

**Pull request merged**